### DISCUSSION TITLE
How to emit events from actor logic such as fromObservable

### DISCUSSION THREAD

**Discussion by reramjiawan** [statelyai/xstate#4968] [Q&A] [answered]
Hi,

I have a question regarding emitting from actorLogic which is invoked such as `fromObservable` or `fromPromise`. When I emit using the provided emit method from the actorLogic I don't understand why this event is not bubbling up to the parent when listening for events.

Example: 

```
const subject = new Subject();

const observableLogic = fromObservable<unknown, any, AnyEventObject>(({ emit }) => {
  return subject.pipe(
    tap((v) => {
      emit({ type: 'EMITTED', value: v });
    })
  );
});

const parentMachine = createMachine({
  initial: 'idle',
  states: {
    idle: {
      invoke: {
        src: observableLogic,
        onDone: {
          target: 'done',
        },
      },
      on: {
        EMITTED: {
          actions: emit({
            type: 'test',
          }),
        },
      },
    },
    done: {},
  },
});

const service = createActor(parentMachine);

service.on('*', (emitted) => {
  console.log('emitted?');
});

service.on('EMITTED', (emitted) => {
  console.log('emitted?');
});

service.start();

// Emit a value from the Subject
subject.next('Hello, world!');
subject.complete();
```

Please let me know if I am doing something wrong here or my expectations are not right.

**Comment by reramjiawan** [comment]
After playing around with XState I have found better suitable patterns. Thanks!

**Reply by GabrielAPineiro** [reply]
If you don't mind me asking, what was your solution for this case? I'm having the same issue, and I'm not understanding why the event is not being called. I have another machine with a similar case and the event is being fired with no problems

**Reply by reramjiawan** [reply, marked as answer]
Hi @GabrielAPineiro,

See this part in the docs https://stately.ai/docs/invoke#invoking-event-observables. Just return events from your observables and the rest is handled by the machine. From there you can do whatever you need, sendTo or whatever else.

**Reply by GabrielAPineiro** [reply]
That's crazy. I have a machine which worked with fromObservable, same logic of actor returning event and it being called automatically. But my second machine only worked with the fromEventObservable.
Thank you for the help!

**Reply by matheo** [reply]
@reramjiawan are you using `fromObservable` in the latest version?
@Andarist I'm streaming EventObjects and trying to use the `emit`ter inside the action logic with no avail.

the workaround I found so far is this one:
```
              onSnapshot: {
                actions: enqueueActions(({ enqueue, event }) => {
                  if (event.snapshot.context) { // omit the first empty snapshot
                    enqueue.raise(event.snapshot.context as any);
                  }
                }),
              },
```
but this is not the idea of the library right?
it's NOT easy :sweat_smile:

**Comment by matheo** [comment]
I see Event observables working, but I don't see `event.output` with `fromObservable`.

I'm doing an API call with Angular rxjs but the result is not propagated to `onDone` nor `onSnapshot`.
The documentation mentions that the data is saved to the `fromObservable` actor context, but how to access it?

PD. I had a target inside `onSnapshot` so I wasn't able to see the event data in the second call.
I expected no empty snapshot when the actor is invoked, only when the Observable emitted tho.

**Reply by davidkpiano** [reply]
Can you please share the code? The result should be present in `onSnapshot`.

> The documentation mentions that the data is saved to the fromObservable actor context, but how to access it?

```ts
actor.getSnapshot().context
```
